src/AbstractSQLCompiler.ts

```diff
@@ -92,7 +92,17 @@ export type NumberTypeNodes =
 	| UnknownTypeNodes;
 
 export type FieldNode = ['Field', string];
-export type ReferencedFieldNode = ['ReferencedField', string, string];
+export type ReferencedFieldNode =
+	| ['ReferencedField', string, string]
+	| [
+			FunctionWrapper,
+			FunctionWrapperParameters,
+			['ReferencedField', string, string],
+	  ];
+export type FunctionWrapper = 'DateTrunc' | string;
+export type FunctionWrapperParameters =
+	| ['EmbeddedText' | string]
+	| [string, string];
```

Comment:
My idea was to have the ReferencedFieldNode Type to be able to have a functionWrapper in terms of a more generic approach, instead of only allowing the DateTrunc. 
Typing it strictly only for the DateTrunc is fine for me. Not typing it to the ReferenceFieldNode type will fail here:  https://github.com/balena-io-modules/odata-to-abstract-sql/pull/73/files#R883 . Then I'd have to manually specify the return Type on the Method in src/odata-to-abstract-sql.ts 
As ReferencedFieldNode is not a subtype of DateNode I don't see the solution.
I don't think the return type should be typed on the function, rather than specifying it in the ReferenceFieldNode, here.

```
export type ReferencedFieldPlainNode = ['ReferencedField', string, string]
export type DateTruncNode = ['DateTrunc', TextTypeNodes, ReferencedFieldPlainNode];
export ReferencedFieldNode = ReferencedFieldPlainNode | DateTruncNode
```

Reply:
So a `ReferencedFieldNode` is specifically one that references a field and a `DateTrunc` node does not directly reference a field, eg imagine you did `UPDATE table t SET date_trunc(t.field) = 'xyz'`, it won't work as you specifically need a field reference and not some other value. That's why there's the types like `DateTypeNodes`/`TextTypeNodes`/etc where all you care about is a certain type of value and not how it comes about.  
In this case the `ReferencedField` of odata-to-abstract-sql should have the return type `ReferencedFieldNode | DateTruncNode` and then if it causes any type issues downstream where a `DateTruncNode` isn't valid then that's just highlighting bugs we need to fix (whether it be a purely typing issue or a runtime issue like the `SET` case would be)

Reply:
Thanks for the feedback, I'm still confused. When I'm not giving the ReferencedFieldNode in the DateTruncNode how will this be resolved?
With
```
export type ReferencedFieldNode = ['ReferencedField', string, string];
export type DateTruncNode = ['DateTrunc', TextTypeNodes, DateTypeNodes];
export type DateTypeNodes = DateNode | DateTruncNode;
```
`export type DateTruncNode = ['DateTrunc', TextTypeNodes, DateTypeNodes];` is missing the link to ReferencedFieldNode.
My understanding is, that the abstractSQL will describe any ReferencedField as a ReferencedFieldNode. And the abstractSQL compiler will resolve this to the actual "table.column" string. The field reference is needed for the inner part of date_trunc as there is no resolver for the "table.field" before. I actually don't see how to circumvent having ReferencedFieldNode as a subtype in the DateTypeNodes as:

```
export type ReferencedFieldNode = ['ReferencedField', string, string];
export type DateTruncNode = ['DateTrunc', TextTypeNodes, DateTypeNodes];
export type DateTypeNodes = DateNode | ReferencedFieldNode;
```

Reply:
Never mind, I've been confusing things. Finally found an approach, which is handling the issue about the ReferencedFields in the odata-to-abstractsql.